M2/Macaulay2/packages/Macaulay2Doc/doc_iterators.m2

```diff
@@ -1,92 +1,128 @@
 doc ///
   Key
-    "iterators"
+    Iterator
   Headline
-    an overview of iterators in Macaulay2
+    class for iterators
   Description
     Text
-      A class in Macaulay2 is @EM "iterable"@ if there is a method
-      function installed for @TO iterator@ that takes as input instances
-      of that class and returns instances of an @EM "iterator"@ class, which
-      has method functions installed for @TO iterator@ (typically @TO identity@)
-      and @TO next@.  Note that iterators are themselves iterable.
+      This is a class designed to simplify writing @TO iterator@ methods.
+      Each instance is a nullary @TO FunctionClosure@ that serves as the
+      @TO next@ method for the iterator.
     Example
-      PrimeIterator = new SelfInitializingType of MutableHashTable;
-      iterator PrimeIterator := identity;
-      next PrimeIterator := p -> p#"val" = nextPrime(p#"val" + 1);
-      i = PrimeIterator {"val" => 1};
-      next i
-      next i
-      next i
+      iter = iterator {1, 2, 3}
+      code iter
     Text
-      If we wish to iterate over an object more than once, then it should have
-      a separate corresponding iterator class.
+      Each call of @TT "next iter"@ is equivalent to @TT "iter()"@.
     Example
-      PrimeIterable = new Type of HashTable;
-      iterator PrimeIterable := x -> PrimeIterator {"val" => 1};
-      P = new PrimeIterable;
-      i = iterator P;
-      next i
-      next i
-      j = iterator P;
-      next j
-      next j
+      next iter
+      iter()
+      next iter
+      iter()
     Text
-      If we would like iteration to stop at a certain point, then the @TO next@
-      method should return the symbol @TO StopIteration@.
+      For example, let us create a class that we may use to iterate over the
+      Fibonacci numbers.
     Example
-      BoundedPrimeIterator = new SelfInitializingType of MutableHashTable;
-      iterator BoundedPrimeIterator := identity;
-      next BoundedPrimeIterator := i -> (
-	  q := nextPrime(i#"val" + 1);
-	  if q > i#"max" then StopIteration else i#"val" = q);
-      BoundedPrimeIterable = new SelfInitializingType of HashTable;
-      iterator BoundedPrimeIterable := x -> BoundedPrimeIterator {
-	  "val" => 1, "max" => x#"max"};
-      P = BoundedPrimeIterable {"max" => 6};
-      i = iterator P;
-      next i
-      next i
-      next i
-      next i
-    Text
-      When an iterator eventually returns @TO StopIteration@, then it (or the
-      iterable to which it corresponds) may be used with @TO toList@, @TO scan@,
-      and @TO "for"@.
-    Example
-      toList P
-      scan(P, print)
-      for x in P list x + 2
+      FibonacciNumbers = new Type of HashTable;
+      iterator FibonacciNumbers := fib -> Iterator(
+	  a := 0;
+	  b := 1;
+	  () -> (
+	      r := a;
+	      (a, b) = (b, a + b);
+	      r));
+      fibonacci = new FibonacciNumbers;
```

Comment:
What would `new Iterator of FibonacciNumbers` do here?

Reply:
```m2
i3 : new Iterator of FibonacciNumbers
stdio:3:1:(3): error: basic type for 'new' method should have been BasicList, HashTable, or Dictionary
```

What behavior were you expecting?

---

I'm thinking that I chose a bad example again...  There's no effective difference between different instances of the `FibonacciNumbers` class, so it's not really a good example of a class that a package author might write and want to install an `iterator` method for.

Maybe something like the following instead?

```m2
i1 : LucasSequence = new Type of HashTable;

i2 : iterator LucasSequence := seq -> Iterator(
         a := 0;
         b := 1;
         () -> (
     	r := a;
     	(a, b) = (b, seq#"P" * a - seq#"Q" * b);
     	r));

i3 : fibonacci = new LucasSequence from {"P" => 1, "Q" => -1};

i4 : for i in fibonacci list if i > 100 then break else i

o4 = {0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89}

o4 : List
```
Or I could just drop the new class altogether from the documentation -- there's no reason that an `Iterator` object has to be returned from some installed `iterator` method.  Perhaps just simplify it down to the following?

```m2
i1 : fibonacci = Iterator(
         a := 0;
         b := 1;
         () -> (
     	r := a;
     	(a, b) = (b, a + b);
     	r));

i2 : for i in fibonacci list if i > 100 then break else i

o2 = {0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89}
```